File: 2016/06.exs
input = with { :ok, f} = File.open( "06_input" ),
             lines = IO.read( f, :all ),
             :ok = File.close( f )
        do
             lines
             |> String.split( "\n", trim: true )
        end

defmodule One do

  def run( input ) do
    input
    |> Enum.map( fn line -> line |> String.graphemes |> Enum.with_index end )
    |> step( [] )
  end

  def format( x, results \\ [] )
  def format( [], results ), do: Enum.join( results, "" )
  def format( [ h | t ], results ) do
    { res, _ } = h
    |> Map.to_list
    |> Enum.sort( &sort/2 )
    |> List.first

    format( t, results ++ [res] )
  end

  def sort( { p_ltr, p_count }, { ltr, count } ) when p_count === count do
    p_ltr < ltr
  end

  def sort( { _, p_count }, { _, count } ) do
    p_count > count
  end

  def step( [], mem ) do
    format( mem )
  end

  def step( list, [] ) do
    [ h | _ ] = list

    mem = 1..Enum.count( h ) |> Enum.map( fn _ -> %{} end )
    
    step( list, mem )
  end

  def step( [ h | t ], mem ) do
    mem = Enum.reduce( h, mem, fn
                   entry, acc -> update( acc, entry )
    end )

    step( t, mem )
  end

  def update( mem, { ltr, idx } ) do
    slot = Enum.at( mem, idx )

    { _, slot } = Map.get_and_update( slot, ltr, fn
                              nil -> { nil, 1 }
                              count -> { count, count + 1 }
    end )

    List.replace_at( mem, idx, slot )
  end

end

input
|> One.run
|> IO.inspect

File: 2017/06.js
const input = [0, 5, 10, 0, 11, 14, 13, 4, 11, 8, 8, 7, 1, 4, 12, 11]

function redistribute(array, source) {
  const size = array.length
  let count = array[source]
  array[source++] = 0

  while (count--) {
    if (source == size) {
      source = 0
    }
    array[source++]++
  }

  return count
}

function maxIndex(array) {
  let i = 0
  let max = 0
  array.forEach((ele, idx) => {
    if (ele > max) {
      i = idx
      max = ele
    }
  })

  return i
}

function step(array) {
  let max = maxIndex(array)
  redistribute(array, max)
  return array.join(' ')
}

function run(array) {
  const cache = {}
  let count = 0
  let key

  while (!(key in cache)) {
    cache[key] = count
    key = step(array)
    count++
  }

  return [count, count - cache[key]]
}

console.log(run(input))
